From 0eeff296ea995113b98f11ea92df696783932dba Mon Sep 17 00:00:00 2001
From: PyScotch <pyscotch@noreply.invalid>
Date: Tue, 13 Oct 2026 13:52:45 +0000
Subject: [PATCH 093/240] Bugfix: do not split contexts in nested dissection
 ordering when deterministic behavior is wanted [report PyScotch]

`hgraphOrderNd()` and `hmeshOrderNd()` split the execution context in
two whenever at least two threads are available. The second
sub-context receives a new pseudo-random generator, so that the
pseudo-random sequences consumed by the recursion depend on the number
of threads, and orderings differ between runs with 1, 2 or 4 threads
even when the deterministic option is set, while the documentation
states that this option yields the same results for any number of
threads.
Run the two sub-tasks in sequence, within the same context, when
deterministic behavior is wanted, as `graphMatchInit()` does for
threaded matching and as `kgraphMapRbPart2()` does for recursive
bipartitioning.

Refs: PS2-G08-2
---
 src/libscotch/hgraph_order_nd.c | 12 +++++++++---
 src/libscotch/hmesh_order_nd.c  | 12 +++++++++---
 2 files changed, 18 insertions(+), 6 deletions(-)

diff --git a/src/libscotch/hgraph_order_nd.c b/src/libscotch/hgraph_order_nd.c
index 357ab36..c946827 100644
--- a/src/libscotch/hgraph_order_nd.c
+++ b/src/libscotch/hgraph_order_nd.c
@@ -1,4 +1,4 @@
-/* Copyright 2004,2007,2010,2012,2014,2016,2018,2019,2021,2023-2025 IPB, Universite de Bordeaux, INRIA & CNRS
+/* Copyright 2004,2007,2010,2012,2014,2016,2018,2019,2021,2023-2026 IPB, Universite de Bordeaux, INRIA & CNRS
 **
 ** This file is part of the Scotch software package for static mapping,
 ** graph partitioning and sparse matrix ordering.
@@ -53,7 +53,7 @@
 /**                # Version 6.1  : from : 01 nov 2021     **/
 /**                                 to   : 21 nov 2021     **/
 /**                # Version 7.0  : from : 05 may 2019     **/
-/**                                 to   : 03 oct 2025     **/
+/**                                 to   : 13 oct 2026     **/
 /**                                                        **/
 /************************************************************/
 
@@ -65,6 +65,7 @@
 
 #include "module.h"
 #include "common.h"
+#include "context.h"
 #include "parser.h"
 #include "graph.h"
 #include "order.h"
@@ -285,6 +286,9 @@ const HgraphOrderNdParam * restrict const paraptr)
   }
   if (o == 0) {
     HgraphOrderNdSplit  spltdat;                  /* Parameters for context splitting */
+#ifndef HGRAPHORDERNDNOTHREAD
+    INT                 deteval;                  /* Flag set if deterministic behavior wanted */
+#endif /* HGRAPHORDERNDNOTHREAD */
 
     spltdat.splttab[0].vnumnbr = vsplisttab[0].vnumnbr;
     spltdat.splttab[0].vnumtab = vsplisttab[0].vnumtab;
@@ -302,7 +306,9 @@ const HgraphOrderNdParam * restrict const paraptr)
     spltdat.revaptr = &o;
 
 #ifndef HGRAPHORDERNDNOTHREAD
-    if (contextThreadLaunchSplit (grafptr->contptr, (ContextSplitFunc) hgraphOrderNd2, &spltdat) != 0) /* If counld not split context to run concurrently */
+    contextValuesGetInt (grafptr->contptr, CONTEXTOPTIONNUMDETERMINISTIC, &deteval);
+    if ((deteval != 0) ||                         /* If deterministic behavior wanted, or if could not split context to run concurrently */
+        (contextThreadLaunchSplit (grafptr->contptr, (ContextSplitFunc) hgraphOrderNd2, &spltdat) != 0))
 #endif /* HGRAPHORDERNDNOTHREAD */
     {
       hgraphOrderNd2 (grafptr->contptr, 0, &spltdat); /* Run tasks in sequence */
diff --git a/src/libscotch/hmesh_order_nd.c b/src/libscotch/hmesh_order_nd.c
index 5f5762b..15c1fe9 100644
--- a/src/libscotch/hmesh_order_nd.c
+++ b/src/libscotch/hmesh_order_nd.c
@@ -1,4 +1,4 @@
-/* Copyright 2004,2007,2008,2018-2021,2023,2024 IPB, Universite de Bordeaux, INRIA & CNRS
+/* Copyright 2004,2007,2008,2018-2021,2023,2024,2026 IPB, Universite de Bordeaux, INRIA & CNRS
 **
 ** This file is part of the Scotch software package for static mapping,
 ** graph partitioning and sparse matrix ordering.
@@ -47,7 +47,7 @@
 /**                # Version 6.0  : from : 15 may 2018     **/
 /**                                 to   : 15 may 2018     **/
 /**                # Version 7.0  : from : 12 sep 2019     **/
-/**                                 to   : 11 sep 2024     **/
+/**                                 to   : 13 oct 2026     **/
 /**                                                        **/
 /************************************************************/
 
@@ -59,6 +59,7 @@
 
 #include "module.h"
 #include "common.h"
+#include "context.h"
 #include "parser.h"
 #include "graph.h"
 #include "order.h"
@@ -241,6 +242,9 @@ const HmeshOrderNdParam * const   paraptr)
   }
   if (o == 0) {
     HmeshOrderNdSplit   spltdat;                  /* Parameters for context splitting */
+#ifndef HMESHORDERNDNOTHREAD
+    INT                 deteval;                  /* Flag set if deterministic behavior wanted */
+#endif /* HMESHORDERNDNOTHREAD */
 
     spltdat.splttab[0].velmnbr = nspmeshdat.ecmpsize[0];
     spltdat.splttab[0].vnodnbr = nspmeshdat.ncmpsize[0];
@@ -258,7 +262,9 @@ const HmeshOrderNdParam * const   paraptr)
     spltdat.revaptr = &o;
 
 #ifndef HMESHORDERNDNOTHREAD
-    if (contextThreadLaunchSplit (meshptr->contptr, (ContextSplitFunc) hmeshOrderNd2, &spltdat) != 0) /* If counld not split context to run concurrently */
+    contextValuesGetInt (meshptr->contptr, CONTEXTOPTIONNUMDETERMINISTIC, &deteval);
+    if ((deteval != 0) ||                         /* If deterministic behavior wanted, or if could not split context to run concurrently */
+        (contextThreadLaunchSplit (meshptr->contptr, (ContextSplitFunc) hmeshOrderNd2, &spltdat) != 0))
 #endif /* HMESHORDERNDNOTHREAD */
     {
       hmeshOrderNd2 (meshptr->contptr, 0, &spltdat); /* Run tasks in sequence */
-- 
2.47.3

